File: broker/src/main/scala/io/octopus/scala/broker/mqtt/server/listener/BindLocalPortListener.scala
package io.octopus.scala.broker.mqtt.server.listener

import io.netty.channel.{ChannelFuture, ChannelFutureListener}
import io.octopus.scala.broker.mqtt.server.TransportBootstrap
import org.slf4j.LoggerFactory

import java.net.InetSocketAddress

/**
 * @author chenxu
 * @date 2022/1/6 7:07 下午
 * @version 1
 */
/**
 * LocalPortReaderFutureListener
 *
 * @param transportName transportName eg：tcp udp wss
 */
class BindLocalPortListener(transportName: String, var ports: Map[String, Int]) extends ChannelFutureListener {

  private val logger = LoggerFactory.getLogger(classOf[TransportBootstrap])

  override def operationComplete(f: ChannelFuture): Unit = {
    if (f.isSuccess) {
      val localAddress = f.channel().localAddress()
      localAddress match {
        case inetAddress: InetSocketAddress =>
          logger.debug("bound {} port: {}", transportName, inetAddress.getPort)
          val port = inetAddress.getPort
          ports += (transportName -> port)
        case _ =>
      }
    }
  }
}

File: broker-kernel/src/main/java/io/octopus/kernel/kernel/listener/BindLocalPortListener.java
package io.octopus.kernel.kernel.listener;

import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelFutureListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetSocketAddress;
import java.util.Map;

/**
 * @author chenxu
 * @version 1
 * @date 2022/7/12 17:14
 */
public class BindLocalPortListener implements ChannelFutureListener {
    private final static Logger LOGGER = LoggerFactory.getLogger(BindLocalPortListener.class);

    private final String transportName;
    private final Map<String, Integer> ports;


    public BindLocalPortListener(String transportName, Map<String, Integer> ports) {
        this.transportName = transportName;
        this.ports = ports;
    }

    @Override
    public void operationComplete(ChannelFuture future) throws Exception {
        if (future.isSuccess()) {
            InetSocketAddress localAddress = (InetSocketAddress) future.channel().localAddress();
            Integer port = localAddress.getPort();
            LOGGER.debug("bound {} port: {}", transportName, port);
            ports.put(transportName, port);
        }
    }
}
